--- mysql-5.6.20/scripts/mysql_install_db.pl.in	2014-07-18 10:48:39.000000000 -0500
+++ mysql-5.6.20/scripts/mysql_install_db.pl.in.new	2014-07-31 16:29:49.337603949 -0500
@@ -113,6 +113,7 @@
   print <<EOF3;
   --rpm                For internal use.  This option is used by RPM files
                        during the MySQL installation process.
+                       Implies --keep-my-cnf option.
   --skip-name-resolve  Use IP addresses rather than hostnames when creating
                        grant table entries.  This option can be useful if
                        your DNS does not work.
@@ -439,7 +440,7 @@
 {
   error($opt,"Specify either --basedir or --srcdir, not both");
 }
-if ( $opt->{'keep-my-cnf'} )
+if ( $opt->{rpm} || $opt->{'keep-my-cnf'} )
 {
   $keep_my_cnf = 1;
 }
@@ -664,7 +665,7 @@
 }
 
 # ----------------------------------------------------------------------
-# Create database directories mysql & test
+# Create database directory mysql
 # ----------------------------------------------------------------------
 
 # FIXME The shell variant uses "mkdir -p":
@@ -697,7 +698,7 @@
   }
 }
 
-foreach my $dir ( $opt->{ldata}, "$opt->{ldata}/mysql", "$opt->{ldata}/test" )
+foreach my $dir ( $opt->{ldata}, "$opt->{ldata}/mysql")
 {
   mkdir($dir, 0700) unless -d $dir;
   if ($opt_user and -w "/")
